File: emisor/algoritmos/contrato.py
"""Contrato que debe cumplir todo algoritmo de integridad (corrección o detección)."""

from dataclasses import dataclass
from enum import Enum
from typing import Protocol


class EstadoVerificacion(Enum):
    SIN_ERROR = "sin_error"
    CORREGIDO = "corregido"
    ERROR_NO_CORREGIBLE = "error_no_corregible"


@dataclass(frozen=True)
class TramaCodificada:
    """Resultado de calcular(): trama completa (datos + redundancia) y parametros para el receptor."""
    bits: str
    parametros: dict


@dataclass(frozen=True)
class ResultadoVerificacion:
    """Resultado de verificar(): estado, bits de datos recuperados (o None) y detalle legible."""
    estado: EstadoVerificacion
    bits: str | None
    detalle: str


class AlgoritmoIntegridad(Protocol):
    nombre: str

    def calcular(self, bits: str) -> TramaCodificada: ...

    def verificar(self, trama: str, parametros: dict) -> ResultadoVerificacion: ...

File: receptor/algoritmos/contrato.js
/**
 * Contrato que debe cumplir todo algoritmo de integridad (corrección o detección).
 * Espejo exacto de emisor/algoritmos/contrato.py: mismos estados y mismas llaves.
 */

export const EstadoVerificacion = Object.freeze({
  SIN_ERROR: "sin_error",
  CORREGIDO: "corregido",
  ERROR_NO_CORREGIBLE: "error_no_corregible",
});

/**
 * @typedef {Object} TramaCodificada
 * @property {string} bits - trama completa: datos + redundancia
 * @property {Object} parametros - info que el receptor necesita para decodificar
 */

/**
 * @typedef {Object} ResultadoVerificacion
 * @property {string} estado - uno de EstadoVerificacion
 * @property {string|null} bits - bits de datos originales, o null si no se pudo recuperar
 * @property {string} detalle - texto legible
 */

/**
 * Interfaz que debe implementar cada algoritmo: { nombre, calcular(bits), verificar(trama, parametros) }.
 */
export class AlgoritmoIntegridad {
  /** @param {string} bits @returns {TramaCodificada} */
  calcular(bits) {
    throw new Error("no implementado");
  }

  /** @param {string} trama @param {Object} parametros @returns {ResultadoVerificacion} */
  verificar(trama, parametros) {
    throw new Error("no implementado");
  }
}
